cachi2/core/models/sbom.py

```diff
@@ -93,3 +93,133 @@ class Sbom(pydantic.BaseModel):
+
+    def __hash__(self) -> int:
+        return hash((self.referenceLocator, self.referenceType, self.referenceCategory))
+
+
+class SPDXPackage(pydantic.BaseModel):
+    """SPDX Package.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/
+    """
+
+    name: str
+    version: Optional[str] = None
+    externalRefs: list[SPDXPackageExternalRef] = []
+
+    def key(self) -> tuple[str, ...]:
+        """Uniquely identifies a package.
+
+        Used mainly for sorting and deduplication.
+        """
+        purls = []
+        for ref in self.externalRefs:
+            if ref.referenceType == "purl":
+                purls.append(ref.referenceLocator)
+        return tuple(purls)
+
+    @classmethod
+    def from_package_dict(cls, package: dict[str, Any]) -> "SPDXPackage":
+        """Create a SPDXPackage from a Cachi2 package dictionary."""
+        if package.get("externalRefs", None):
+            external_refs = [
+                SPDXPackageExternalRef(
+                    referenceLocator=er["referenceLocator"],
+                    referenceType=er["referenceType"],
+                    referenceCategory=er["referenceCategory"],
+                )
+                for er in package["externalRefs"]
+            ]
+        else:
+            external_refs = []
+
+        return SPDXPackage(
+            name=package.get("name", None),
+            version=package.get("version", None),
+            externalRefs=external_refs,
+        )
+
+
+class SPDXCreationInfo(pydantic.BaseModel):
+    """SPDX Creation Information.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/document-creation-information/
+    """
+
+    creators: list[str] = []
+
+
+def deduplicate_spdx_packages(items: Iterable[SPDXPackage]) -> list[SPDXPackage]:
+    """ Deduplicate SPDX packages and merge external references.
+
+    If package with same name and version is found multiple times in the list,
+    merge external references of all the packages into one package.
+    """
+    unique_items = {}
+    for item in items:
+        key = (item.name, item.version)
+        if key not in unique_items:
+            unique_items[key] = SPDXPackage(name=item.name, version=item.version)
+            unique_items[key].externalRefs = item.externalRefs[:]
+        else:
+            unique_items[key].externalRefs.extend(item.externalRefs)
+
+    for item in unique_items.values():
+        item.externalRefs = sorted(
+            list(set(item.externalRefs)),
+            key=lambda ref: (ref.referenceLocator, ref.referenceType, ref.referenceCategory),
+        )
+
+    return sorted(list(unique_items.values()), key=lambda item: (item.name, item.version))
+
+
+class SPDXSbom(pydantic.BaseModel):
+    """Software bill of materials in the SPDX format.
+
+    See full specification at:
+    https://spdx.github.io/spdx-spec/v2.3
+    """
+
+    spdxVersion: str = "SPDX-2.3"
```

Comment:
Would a `Literal` better convey the intent here?

Reply:
yeah, makes sense